[NNC] Complete SimpleIREvaluator support for bitwise ops (#48053)

penguinwu opened the pull request:
Summary: Add missing types for bitwise_ops in `SimpleIREvaluator`

This is the first part of fixes for issue #48053.
- Original implementation of bitwise_ops supports only int operands, the
fix all support for integral types supported by the IR

Test Plan: `python test/test_jit_fuser_te.py TestTEFuser.test_bitwise_ops`

Reviewers: 

Subscribers:

Tasks: 48053

Tags: NNC

Commit:
[NNC] Complete SimpleIREvaluator support for bitwise ops (#48053)

Summary: Add missing types for bitwise_ops in SimpleIREvaluator

Original implementation of bitwise_ops supports only int operands, the
fix all support for integral types supported by the IR

Test Plan: python test/test_jit_fuser_te.py TestTEFuser.test_bitwise_ops

Reviewers: eellison

Subscribers:

Tasks: 48053

Tags: NNC

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.h:
```diff
@@ -368,11 +369,21 @@ class SimpleIREvaluator : public CodeGen, public IRVisitor {
+    break;
+        AT_FORALL_INT_TYPES(TYPE_CASE);
+#undef TYPE_CASE
+        default:
```
Do we need to handle `ScalarType::Bool` here as well?

penguinwu replied:
You are right. According to `ir.h`, `kAnd`, `kOr`, `kXor` applies to all integral types (including `Bool`, according to `tensorexpr::is_integral()`):

```
class And : public BinaryOpNode<And> {
 public:
  And(const Expr* lhs, const Expr* rhs)
      : BinaryOpNode(lhs, rhs, IRNodeType::kAnd) {
    if (!lhs->dtype().is_integral()) {
      throw unsupported_dtype();
    }
    if (lhs->dtype() != rhs->dtype()) {
      throw malformed_input("bad dtype in And");
    }
  }
};
``` 

But for `kLshift` and `kRshift` (according to `tensorexpr/ir.h`), the supported type seems to be only `ScalarType::Int`
```
class Lshift : public BinaryOpNode<Lshift> {
 public:
  Lshift(const Expr* lhs, const Expr* rhs)
      : BinaryOpNode(lhs, rhs, IRNodeType::kLshift) {
    if (lhs->dtype().scalar_type() != ScalarType::Int) {
      throw unsupported_dtype();
    }
    if (lhs->dtype() != rhs->dtype()) {
      throw malformed_input("bad dtype in Lshift");
    }
  }
};
```

I'll make the changes accordingly.

ZolotukhinM reviewed (approved):
Good work, thanks for fixing this! I have a question about handling bool types, but apart from that the PR looks good!

penguinwu commented:
@ZolotukhinM BTW, I could not make sense out of the CI failure. It seems to be a build issue. What should I do about it?

```
🕵️ 1 new failure recognized by patterns

The following CI failures do not appear to be due to upstream breakages:

See CircleCI build binary_linux_libtorch_3_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (1/1)

Step: "Build" (full log | diagnosis details | 🔁 rerun)

 RuntimeError: no cmake or cmake3 with version >= 3.5.0 found
```

ZolotukhinM commented:
> @ZolotukhinM BTW, I could not make sense out of the CI failure. It seems to be a build issue. What should I do about it?

Yes, unfortunately this happens. It doesn't look like your change introduced the failure, so we could do either of two things: 1) rebase your PR retriggering the CI in a hope that the issue disappears, 2) ignore it (it's really sad that we might have to do this, but if your PR is clearly unrelated, it's ok).

Commit:
[NNC] Add bool support for and/or/xor ops ((#48053)

Summary:

- Add support for bool type for an/or/xor
- Restrict implementation of shift left/right to int type
  (to be consistent w/ IR specification)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags: bootcamp

penguinwu commented:
Rebased and addressed the review comment on `bool` type. Please review again.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.h:
```diff
@@ -368,11 +387,31 @@ class SimpleIREvaluator : public CodeGen, public IRVisitor {
+
+    if (expr_type == IRNodeType::kLshift || expr_type == IRNodeType::kRshift) {
+      switch (lhs_v.dtype().scalar_type()) {
+        case ScalarType::Int:
```
But we now need to support different dtypes here as well rather than always using the `int` version, right? I think we need to use `AT_FORALL_INT_TYPES` here as well.

Btw, we don't need to add `_AND` to the macro if the only reason for that is to support the bool case in the bitwise ops - we can add an explicit `case` for `ScalarType::bool` instead.

penguinwu replied:
in `ir.h`, the constructor of `Lshift` IR seems to only support operands of type `ScalarType::Int`. 

```
class Lshift : public BinaryOpNode<Lshift> {
 public:
  Lshift(const Expr* lhs, const Expr* rhs)
      : BinaryOpNode(lhs, rhs, IRNodeType::kLshift) {
    if (lhs->dtype().scalar_type() != ScalarType::Int) {
      throw unsupported_dtype();
    }
    if (lhs->dtype() != rhs->dtype()) {
      throw malformed_input("bad dtype in Lshift");
    }
  }
};
```

I was trying to make the implementation of `Lshift` and `Rshift` to be consistent w/ its IR definition. But supporting other integer types (which also makes sense for these IRs) in the implementation will not cause any trouble. So I am fine either way. It's your call.

As for `ScalarType::bool`, I will add an explicit case switch for it to make it more readable.

ZolotukhinM replied:
> I was trying to make the implementation of Lshift and Rshift to be consistent w/ its IR definition. But supporting other integer types (which also makes sense for these IRs) in the implementation will not cause any trouble. So I am fine either way. It's your call.

Ultimately the ground truth here is the semantics of aten ops that we're representing in tensor expressions. In case of shifts the following script could serve as a reference for us:
```
import torch

dtypes = [
  torch.uint8,
  torch.int8,
  torch.int16,
  torch.int32,
  torch.float16,
  torch.float32,
  torch.float64,
  torch.bool,
]

for dev in ['cpu', 'cuda']:
    for dt in dtypes:
        x = torch.ones(4, dtype=dt, device=dev)
        try:
            y = x << 2
            print(y)
        except Exception as e:
            print(e)
```

If we run it, we'll see that the dtypes should be preserved through these operations:
```
tensor([4, 4, 4, 4], dtype=torch.uint8)
tensor([4, 4, 4, 4], dtype=torch.int8)
tensor([4, 4, 4, 4], dtype=torch.int16)
tensor([4, 4, 4, 4], dtype=torch.int32)
"lshift_cpu" not implemented for 'Half'
tensor([4., 4., 4., 4.])
tensor([4., 4., 4., 4.], dtype=torch.float64)
"lshift_cpu" not implemented for 'Bool'
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.uint8)
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.int8)
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.int16)
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.int32)
tensor([4., 4., 4., 4.], device='cuda:0', dtype=torch.float16)
tensor([4., 4., 4., 4.], device='cuda:0')
tensor([4., 4., 4., 4.], device='cuda:0', dtype=torch.float64)
"lshift_cuda" not implemented for 'Bool'
```

But as you noted, this is a slightly bigger change and seems like a separate issue to me. Please feel free to land this PR and fix behavior of shifts later in a different one.

ZolotukhinM replied:
Btw, I'm adding tests that would expose this discrepancy between the eager and TE-compiled behavior of shift operations in #48292.
